lua/illuminate/providers/treesitter.lua

```diff
@@ -17,29 +17,71 @@ function M.get_references(bufnr)
+    -- If we can get back to this node through `find_usages`,
+    -- then we know that this node was a definition,
+    -- because `find_definition(some_node)` returns `some_node`,
+    -- also when `some_node` is a definition,
+    -- and `find_usages(some_node)` returns an empty list
+    -- if `some_node` isn't an identifier
+    -- (also known from oversvation, in the absense of specification).
+
+    local usages = locals.find_usages(maybe_def_node, scope, bufnr)
+    for _, node in ipairs(usages) do
+        local range = { node:range() }
+        -- If something is found through usages,
+        -- we assume that it is a read.
+        local kind = vim.lsp.protocol.DocumentHighlightKind.Read
+        -- But if it is the `maybe_def_node` that we skipped (did not mark) above
+        -- because we were not sure if it was an identifier,
+        -- we now know that it is an identifier,
+        -- because we found it through `find_usages`.
+        if node == maybe_def_node then
+            -- Yet, we cannot be sure that it is the definition of that identifier,
+            -- because some things have their definition outside of the current file,
+            -- for example, imported APIs.
+            -- Thus, we must also check if this node is a definition.
+            -- Performance: despite being in a loop, this check is done at most once,
+            -- since `usages` contains no duplicates and only one `node` equals `maybe_def_node`.
+           if is_definition(locals, node) then
+               kind = vim.lsp.protocol.DocumentHighlightKind.Write
+           end
+        end
+
         table.insert(refs, {
             { range[1], range[2] },
             { range[3], range[4] },
-            vim.lsp.protocol.DocumentHighlightKind.Write,
+            kind,
         })
     end
 
-    local usages = locals.find_usages(def_node, scope, bufnr)
-    for _, node in ipairs(usages) do
-        if node ~= def_node then
-            local range = { node:range() }
-            table.insert(refs, {
-                { range[1], range[2] },
-                { range[3], range[4] },
-                vim.lsp.protocol.DocumentHighlightKind.Read,
-            })
+    return refs
+end
+
+-- Check if `node` is marked as a defintion (`locals.definition[.something]`)
+-- Specification: https://github.com/nvim-treesitter/nvim-treesitter/blob/51bba660a89e0027929206b622c9c1cbdd995cfb/CONTRIBUTING.md#locals
+function is_definition(locals, node)
+    for _, entry in ipairs(locals.get_definitions(bufnr, 'locals.definition')) do
+        if entry.node  == node then
+            -- node marked as `locals.definition`
+            return true
+        else
+            for _, sub in pairs(entry) do
+                if sub.node == node then
+                    -- node marked as `locals.definition.something`
+                    return true
+                end
+            end
         end
     end
-
-    return refs
+    return false
```

Comment:
This is too brittle for me to merge it.

Reply:
I understand. This was also my least favorite part of the implementation. What would you say if a function with this effect was not implemented here, but in `nvim-treesitter`?
I think such a function is valuable and that they might be interested. If you're in favor, I'd volunteer to bring the issue up with them, to either add such an API or ask how it should be done instead.